subsys/net/l2/canbus/canbus.c

```diff
@@ -1,49 +1,1237 @@
+	}
+
+	ret = process_cf_data(pkt, rx_ctx);
+	if (ret < 0) {
+		return NET_DROP;
+	}
+
+	net_pkt_unref(pkt);
+
+	if (rx_ctx->rem_len == 0) {
+		rx_ctx->state = NET_CAN_RX_STATE_FIN;
+		ret = net_recv_data(pkt->iface, rx_ctx->pkt);
+		if (ret < 0) {
+			NET_ERR("Packet dropped by NET stack");
+			net_pkt_unref(pkt);
+		}
+	} else {
+		z_add_timeout(&rx_ctx->timeout, rx_timeout,
+			      z_ms_to_ticks(NET_CAN_BS_TIME));
+
+		if (NET_CAN_BS != 0 && !mcast) {
+			rx_ctx->act_block_nr++;
+			if (rx_ctx->act_block_nr >= NET_CAN_BS) {
+				NET_DBG("BS reached. Send FC");
+				src.addr = get_src_lladdr(pkt);
+				dest.addr = get_dest_lladdr(pkt);
+				net_can_dev = net_if_get_device(pkt->iface);
+				ret = send_fc(net_can_dev, &src, &dest,
+					NET_CAN_PCI_FS_CTS);
+				if (ret) {
+					NET_ERR("Failed to send FC CTS. BS: %d",
+						NET_CAN_BS);
+					rx_report_err(rx_ctx->pkt);
+					return NET_OK;
+				}
+
+				rx_ctx->act_block_nr = 0;
+			}
+		}
+	}
+
+	return NET_OK;
+}
+
+static enum net_verdict process_ff(struct net_pkt *pkt)
+{
+	struct device *net_can_dev = net_if_get_device(pkt->iface);
+	struct can_net_isotp_rx_ctx *rx_ctx = NULL;
+	struct net_pkt *new_pkt = NULL;
+	int ret;
+	struct net_can_lladdr src, dest;
+	u32_t msg_len;
+	u8_t data_len;
+	bool mcast;
+
+	mcast = dest_is_mcast(pkt);
+	src.addr = get_src_lladdr(pkt);
+	dest.addr = get_dest_lladdr(pkt);
+	net_pkt_cursor_init(pkt);
+
+	msg_len = receive_get_ff_length(pkt);
+	new_pkt = net_pkt_rx_alloc_with_buffer(pkt->iface,
+					       msg_len + 2 * sizeof(struct net_can_lladdr),
+					       AF_INET6, 0,
+					       NET_CAN_ALLOC_TIMEOUT);
+	if (!new_pkt) {
+		NET_ERR("Failed to obtain net_pkt with size of %d",
+			msg_len + 2 * sizeof(struct net_can_lladdr));
+
+		if (!mcast) {
+			send_fc(net_can_dev, &src, &dest, NET_CAN_PCI_FS_OVFLW);
+		}
+
+		goto err;
+	}
+
+	rx_ctx = get_rx_ctx(NET_CAN_RX_STATE_UNUSED, 0);
+	if (!rx_ctx) {
+		NET_ERR("No rx context left");
+
+		if (!mcast) {
+			send_fc(net_can_dev, &src, &dest, NET_CAN_PCI_FS_OVFLW);
+		}
+
+		goto err;
+	}
+
+	rx_ctx->act_block_nr = 0;
+	rx_ctx->pkt = new_pkt;
+	new_pkt->can_rx_ctx = rx_ctx;
+
+	net_pkt_cursor_init(new_pkt);
+	cpy_lladdr(new_pkt, pkt);
+
+	rx_ctx->sn = 1;
+	data_len = net_pkt_remaining_data(pkt);
+	ret = net_pkt_copy(new_pkt, pkt, data_len);
+	if (ret) {
+		NET_ERR("Failed to write to pkt [%d]", ret);
+		goto err;
+	}
+
+	rx_ctx->rem_len = msg_len - data_len;
+	net_pkt_unref(pkt);
+
+	if (!mcast) {
+		/* switch src and dest because we are answering */
+		ret = send_fc(net_can_dev, &src, &dest, NET_CAN_PCI_FS_CTS);
+		if (ret) {
+			NET_ERR("Failed to send FC CTS.");
+			rx_report_err(new_pkt);
+			return NET_OK;
+		}
+	}
+
+	/* At this point we expect to get Consecutive frames directly */
+	z_add_timeout(&rx_ctx->timeout, rx_timeout,
+		      z_ms_to_ticks(NET_CAN_BS_TIME));
+
+	rx_ctx->state = NET_CAN_RX_STATE_CF;
+	if (mcast) {
+		NET_INFO("Processed FF (multicast). Msg length: %u CTX: %p",
```

Comment:
Not sure if it is useful to use the INFO level, it usually just prints too much data. Would NET_DBG() be more proper here?

Reply:
This was for debugging only. Changed it to DBG